File: cpp_drill/Problems/PQ01_TaskScheduler.cpp
#include <iostream>
#include <queue>
#include <unordered_map>
using namespace std;

// REF : https://leetcode.com/problems/task-scheduler/
// NOTES :
//      MaxHeap to keep the count of the each element.
//      Queue to cycle through the different elements which stores
//      (count, idle time) idle time means when this element can be
//      scheduled for job. Whose idle time matches with the current time,
//      pop that element from the Queue and put it in MaxHeap.
//      Take out from Queue ---> Put into MaxHeap

// T: O(n * c * log 26) ~ O(n)
// S: O(1), bcz Queue and maxHeap size can go upto 26 distinct elements

class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {

        unordered_map<char, int> freq;
        for (auto &ch : tasks)
            freq[ch]++;

        priority_queue<int> maxHeap;
        for (auto &i : freq)
            maxHeap.push(i.second);

        deque<pair<int, int>> q;
        int time = 0;
        
        while ( (!maxHeap.empty()) || (!q.empty()) ) {
            time++;
            if (!maxHeap.empty()) {

                int cnt = maxHeap.top(); maxHeap.pop();

                cnt--;          // It's not python, we can have MaxHeap
                if (cnt != 0)
                    q.push_back(make_pair(cnt, time + n));

            }

            if ( (!q.empty()) && (q.front().second == time) ) {
                int new_freq = q.front().first; q.pop_front();
                maxHeap.push(new_freq);
            }
        }
        return time;
    }
};

class Solution2 {
public:
    int leastInterval(vector<char>& tasks, int n) {
        unordered_map<char, int> freq;
        for (auto &i : tasks)
            freq[i]++;

        priority_queue<int> maxHeap;
        for (auto &i : freq)
            maxHeap.push(i.second);

        int time = 0;
        queue<pair<int, int>> q;
        
        while ((!maxHeap.empty()) || (!q.empty())) {
            time++;
            if (!maxHeap.empty()) {
                int cnt = maxHeap.top(); maxHeap.pop();
                cnt--;

                if (cnt)
                    q.push({cnt, time + n});
            }

            if ((!q.empty()) && (q.front().second == time)) {
                int new_freq = q.front().first; q.pop();
                maxHeap.push(new_freq);
            }
        }
        return time;
    }
};

int main() {
    vector<char> tasks = {'A','A','A','B','B','B'};
    int n = 2;
    cout << "output : " << Solution().leastInterval(tasks, n) << endl;
    return 0;
}

File: py_drill/Problems/PQ01_TaskScheduler.py
# REF : https://leetcode.com/problems/task-scheduler/
# NOTES:
#       maxHeap + Queue
#       python has only Min Heap. For Max Heap can be implemented using -ve numbers.

# T: O(n * c * log 26) ~ O(n)
# S: O(1), bcz Queue and maxHeap size can go upto 26 distinct elements

import heapq
from collections import Counter, deque

class Solution:
    def leastInterval(self, tasks: list[str], n: int) -> int:

        freq_dict = Counter(tasks)
        maxHeap = [ -i for i in freq_dict.values()]
        heapq.heapify(maxHeap)
        q = deque()

        time = 0
        while q or maxHeap:
            time += 1                                   # Increase timeline by 1 step by step
            if maxHeap:
                cnt = heapq.heappop(maxHeap)            # pop max freq element from heap
                cnt += 1                                # and reduce freq by 1 and

                if cnt:                                 # if count is non-zero then add it into queue
                    q.append((cnt, time + n))           # if count is 0, means that element has been consumed
                                                        # to its all freqency

            if q and q[0][1] == time:                   # when element's next occurance time matches with time
                new_freq = q.popleft()[0]               # then pop it from queue and push into heap
                heapq.heappush(maxHeap, new_freq)

        return time



def main():
    tasks = ["A","A","A","B","B","B"]
    n = 2
    print("output : {}".format(Solution().leastInterval(tasks, n)))

if __name__ == "__main__":
    main()